// @splitsoftware/splitio/es/settings/node.js
import { settingsValidation } from '@splitsoftware/splitio-commons/esm/utils/settingsValidation';
import { validateLogger } from '@splitsoftware/splitio-commons/esm/utils/settingsValidation/logger/builtinLogger';
import { LocalhostFromFile } from '../sync/offline/LocalhostFromFile';
import { defaults } from './defaults/node';
import { validateStorage } from './storage/node';
import { validateRuntime } from './runtime/node';
var params = {
    defaults: defaults,
    runtime: validateRuntime,
    storage: validateStorage,
    logger: validateLogger,
    localhost: function () { return LocalhostFromFile(); },
    consent: function () { return undefined; }, // resets settings.userConsent to the default
    // In Node.js the SDK ignores `config.integrations`, so a validator for integrations is not required
};
export function settingsFactory(config) {
    var settings = settingsValidation(config, params);
    // if provided, keeps reference to the `requestOptions` object
    if (settings.sync.requestOptions)
        settings.sync.requestOptions = config.sync.requestOptions;
    return settings;
}
